src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs

```diff
@@ -142,37 +155,61 @@ private bool IsTopLevelWildcard(string host)
         }
 
         // This does not duplicate format validations that are expected to be performed by the host.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private bool CheckHost(HttpContext context, IList<StringSegment> allowedHosts)
         {
-            var host = new StringSegment(context.Request.Headers[HeaderNames.Host].ToString()).Trim();
+            var host = context.Request.Headers[HeaderNames.Host].ToString();
```

Comment:
@Tratcher is there a reason to allow whitespace before and after the host value? Kestrel will trim it and I assume IIS will also?

rfc7230  3.2.4.  Field Parsing says https://tools.ietf.org/html/rfc7230#page-25
```
Optional whitespace occurring before the first non-whitespace octet of the field value or after the last
non-whitespace octet of the field value ought to be excluded by
parsers when extracting the field value from a header field
```
I've changed it to reject rather than trim

Reply:
It was not making any assumptions about the underlying server. It was valid input so we tried to allow for it. Does this part really show up in your trace?

Reply:
StringSegment is a 16 byte struct so passed via stack copy
```
Type layout for 'StringSegment'
Size: 16 bytes. Paddings: 0 bytes (%0 of empty space)
|=================================================|
|   0-7: String <Buffer>k__BackingField (8 bytes) |
|-------------------------------------------------|
|  8-11: Int32 <Offset>k__BackingField (4 bytes)  |
|-------------------------------------------------|
| 12-15: Int32 <Length>k__BackingField (4 bytes)  |
|=================================================|
```

`StringSegment.Trim()` => `TrimStart().TrimEnd()`

TrimStart and TrimEnd both fix the string as well as having a loop both of which block inlining; so it gets stack copies for each of them, then stack copied back out for the return; in addition to the fixing cost.

Could change Trim to first check is the start and end are whitespace etc and optimize there; however its either a parser error, or a user error (host set in pipeline with leading/trailing whitespace) and `TryProcessHosts` doesn't trim the hosts in the allow lists so its not a "round-trippable" error as it won't match if used in both places; so seems easiest to reject?

Reply:
> Does this part really show up in your trace?
```
dotnet new web
dotnet publish -c Release
```
```
 wrk -c 128 -t 4 -d 120 -H "Accept: application/json,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7" -H "Connection: keep-alive" http://127.0.0.1:5000/ -s ./wrk/scripts/pipeline.lua -- 16
```
On sampling it CheckHost+EnsureConfigured shows up as >10% of the rest of the pipeline 

CheckHost+EnsureConfigured: 20,569+1198 = 21,767ms
EndpointRoutingMiddleware onwards: 213,871ms

21,767/213,871 * 100 = 10.2%

![image](https://user-images.githubusercontent.com/1142958/56096763-69acd680-5ee4-11e9-8fc7-1efbf04e791d.png)

Reply:
TrimStart and TrimEnd are players here; though the effect of logging is more significant

![image](https://user-images.githubusercontent.com/1142958/56096894-a88f5c00-5ee5-11e9-8609-ead7f9db6f59.png)